File: src/mcp_server/tools.py
"""
Implementaciones puras de las herramientas de archivos del MCP Server.

Todas las funciones son independientes y pueden ser importadas
tanto por el servidor MCP como por el bot de Telegram.
"""
import datetime
import shutil
from pathlib import Path

from .config import FILES_DIR, MAX_FILE_SIZE_MB

# Extensiones de texto que se pueden buscar/leer
TEXT_EXTENSIONS = {
    ".txt", ".md", ".csv", ".json", ".yaml", ".yml",
    ".log", ".ini", ".cfg", ".toml", ".html", ".xml",
    ".rst", ".org", "",
}


def _safe_path(filename: str) -> Path:
    """
    Resuelve la ruta y verifica que esté dentro de FILES_DIR.
    Previene path traversal (ej: '../../etc/passwd').
    """
    if not filename or not filename.strip():
        return FILES_DIR
    resolved = (FILES_DIR / filename.strip()).resolve()
    if not str(resolved).startswith(str(FILES_DIR.resolve())):
        raise ValueError("Acceso denegado: la ruta está fuera del directorio permitido")
    return resolved


# ==============================================================
# Herramienta: read_file
# ==============================================================
def read_file(filename: str) -> str:
    """
    Lee el contenido de un archivo de texto.

    Args:
        filename: Nombre o ruta relativa del archivo a leer.

    Returns:
        Contenido del archivo como texto, o mensaje de error.
    """
    try:
        path = _safe_path(filename)
    except ValueError as e:
        return f"Error: {e}"

    if not path.exists():
        return f"Error: El archivo '{filename}' no existe."
    if not path.is_file():
        return f"Error: '{filename}' es un directorio, no un archivo."

    size_mb = path.stat().st_size / (1024 * 1024)
    if size_mb > MAX_FILE_SIZE_MB:
        return (
            f"Error: Archivo demasiado grande ({size_mb:.1f} MB). "
            f"Máximo permitido: {MAX_FILE_SIZE_MB} MB."
        )

    try:
        return path.read_text(encoding="utf-8")
    except UnicodeDecodeError:
        return f"Error: '{filename}' no parece ser un archivo de texto (posible binario)."
    except Exception as e:
        return f"Error al leer el archivo: {e}"


# ==============================================================
# Herramienta: write_file
# ==============================================================
def write_file(filename: str, content: str) -> str:
    """
    Escribe contenido en un archivo de texto.
    Crea el archivo si no existe; lo sobreescribe si ya existe.

    Args:
        filename: Nombre o ruta relativa del archivo a escribir.
        content:  Contenido a escribir.

    Returns:
        Mensaje de éxito o error.
    """
    try:
        path = _safe_path(filename)
    except ValueError as e:
        return f"Error: {e}"

    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(content, encoding="utf-8")
        size = path.stat().st_size
        return f"Archivo '{filename}' guardado correctamente ({size} bytes, {len(content)} caracteres)."
    except Exception as e:
        return f"Error al escribir el archivo: {e}"


# ==============================================================
# Herramienta: list_files
# ==============================================================
def list_files(directory: str = "") -> str:
    """
    Lista archivos y directorios.

    Args:
        directory: Subdirectorio a listar (vacío = directorio raíz).

    Returns:
        Lista formateada de archivos y directorios.
    """
    try:
        path = _safe_path(directory) if directory else FILES_DIR
    except ValueError as e:
        return f"Error: {e}"

    if not path.exists():
        return f"Error: El directorio '{directory}' no existe."
    if not path.is_dir():
        return f"Error: '{directory}' es un archivo, no un directorio."

    items = []
    for item in sorted(path.iterdir()):
        try:
            rel = item.relative_to(FILES_DIR)
            if item.is_dir():
                count = sum(1 for _ in item.iterdir())
                items.append(f"[DIR]  {rel}/  ({count} elementos)")
            else:
                size = item.stat().st_size
                items.append(f"[FILE] {rel}  ({size:,} bytes)")
        except Exception:
            continue

    if not items:
        return "El directorio está vacío."

    header = f"Contenido de {'/' if not directory else directory}  ({len(items)} elementos):\n"
    return header + "\n".join(items)


# ==============================================================
# Herramienta: delete_file
# ==============================================================
def delete_file(filename: str) -> str:
    """
    Elimina un archivo o directorio.

    Args:
        filename: Nombre o ruta relativa del archivo/directorio a eliminar.

    Returns:
        Mensaje de éxito o error.
    """
    try:
        path = _safe_path(filename)
    except ValueError as e:
        return f"Error: {e}"

    if not path.exists():
        return f"Error: '{filename}' no existe."

    try:
        if path.is_dir():
            shutil.rmtree(path)
            return f"Directorio '{filename}' eliminado correctamente."
        else:
            path.unlink()
            return f"Archivo '{filename}' eliminado correctamente."
    except Exception as e:
        return f"Error al eliminar: {e}"


# ==============================================================
# Herramienta: search_in_files
# ==============================================================
def search_in_files(query: str, directory: str = "") -> str:
    """
    Busca texto en todos los archivos de texto (insensible a mayúsculas).

    Args:
        query:     Texto a buscar.
        directory: Directorio donde buscar (vacío = busca en todos).

    Returns:
        Resultados de la búsqueda con nombre de archivo y líneas coincidentes.
    """
    if not query or not query.strip():
        return "Error: La búsqueda no puede estar vacía."

    try:
        search_dir = _safe_path(directory) if directory else FILES_DIR
    except ValueError as e:
        return f"Error: {e}"

    if not search_dir.exists():
        return f"Error: El directorio de búsqueda no existe."

    results = []
    files_searched = 0
    query_lower = query.lower()

    for path in sorted(search_dir.rglob("*")):
        if not path.is_file():
            continue
        if path.suffix.lower() not in TEXT_EXTENSIONS:
            continue

        try:
            content = path.read_text(encoding="utf-8", errors="ignore")
            files_searched += 1
            lines = content.splitlines()
            matches = [
                (i + 1, line)
                for i, line in enumerate(lines)
                if query_lower in line.lower()
            ]

            if matches:
                rel_path = path.relative_to(FILES_DIR)
                results.append(f"\n=== {rel_path} ===")
                shown = matches[:5]
                for line_num, line in shown:
                    results.append(f"  Línea {line_num}: {line.strip()[:200]}")
                if len(matches) > 5:
                    results.append(f"  ... y {len(matches) - 5} coincidencia(s) más")
        except Exception:
            continue

    if not results:
        return f"No se encontró '{query}' en {files_searched} archivo(s) revisado(s)."

    header = f"Se encontró '{query}' en {len(results)} archivo(s) (de {files_searched} revisados):"
    return header + "".join(results)


# ==============================================================
# Herramienta: get_file_info
# ==============================================================
def get_file_info(filename: str) -> str:
    """
    Obtiene metadatos de un archivo (tamaño, fechas, tipo).

    Args:
        filename: Nombre o ruta relativa del archivo.

    Returns:
        Información del archivo formateada.
    """
    try:
        path = _safe_path(filename)
    except ValueError as e:
        return f"Error: {e}"

    if not path.exists():
        return f"Error: '{filename}' no existe."

    try:
        stat = path.stat()
        modified = datetime.datetime.fromtimestamp(stat.st_mtime).strftime("%Y-%m-%d %H:%M:%S")
        created = datetime.datetime.fromtimestamp(stat.st_ctime).strftime("%Y-%m-%d %H:%M:%S")
        size_kb = stat.st_size / 1024

        lines = [
            f"Nombre    : {path.name}",
            f"Ruta      : {path.relative_to(FILES_DIR)}",
            f"Tipo      : {'Directorio' if path.is_dir() else 'Archivo'}",
            f"Tamaño    : {stat.st_size:,} bytes ({size_kb:.1f} KB)",
            f"Creado    : {created}",
            f"Modificado: {modified}",
        ]
        if path.is_dir():
            count = sum(1 for _ in path.iterdir())
            lines.append(f"Elementos : {count}")
        elif path.suffix:
            lines.append(f"Extensión : {path.suffix}")

        return "\n".join(lines)
    except Exception as e:
        return f"Error al obtener información: {e}"

File: internal/tools/tools.go
package tools

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/erosao/mcp/internal/config"
)

var textExtensions = map[string]bool{
	".txt": true, ".md": true, ".csv": true, ".json": true,
	".yaml": true, ".yml": true, ".log": true, ".ini": true,
	".cfg": true, ".toml": true, ".html": true, ".xml": true,
	".rst": true, ".org": true, "": true,
}

func safePath(filename string) (string, error) {
	filename = strings.TrimSpace(filename)
	filesDir, err := filepath.Abs(config.FilesDir)
	if err != nil {
		return "", err
	}
	if filename == "" {
		return filesDir, nil
	}
	resolved, err := filepath.Abs(filepath.Join(filesDir, filename))
	if err != nil {
		return "", err
	}
	if resolved != filesDir && !strings.HasPrefix(resolved, filesDir+string(filepath.Separator)) {
		return "", fmt.Errorf("acceso denegado: la ruta está fuera del directorio permitido")
	}
	return resolved, nil
}

func ReadFile(filename string) string {
	path, err := safePath(filename)
	if err != nil {
		return "Error: " + err.Error()
	}
	info, err := os.Stat(path)
	if err != nil {
		return fmt.Sprintf("Error: El archivo '%s' no existe.", filename)
	}
	if info.IsDir() {
		return fmt.Sprintf("Error: '%s' es un directorio, no un archivo.", filename)
	}
	if sizeMB := float64(info.Size()) / (1024 * 1024); sizeMB > float64(config.MaxFileSizeMB) {
		return fmt.Sprintf("Error: Archivo demasiado grande (%.1f MB). Máximo: %d MB.", sizeMB, config.MaxFileSizeMB)
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return "Error al leer el archivo: " + err.Error()
	}
	return string(data)
}

func WriteFile(filename, content string) string {
	path, err := safePath(filename)
	if err != nil {
		return "Error: " + err.Error()
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return "Error al crear directorios: " + err.Error()
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		return "Error al escribir el archivo: " + err.Error()
	}
	info, _ := os.Stat(path)
	return fmt.Sprintf("Archivo '%s' guardado correctamente (%d bytes, %d caracteres).",
		filename, info.Size(), len([]rune(content)))
}

func ListFiles(directory string) string {
	var path string
	var err error
	if directory == "" {
		path, err = filepath.Abs(config.FilesDir)
	} else {
		path, err = safePath(directory)
	}
	if err != nil {
		return "Error: " + err.Error()
	}
	info, err := os.Stat(path)
	if err != nil {
		return fmt.Sprintf("Error: El directorio '%s' no existe.", directory)
	}
	if !info.IsDir() {
		return fmt.Sprintf("Error: '%s' es un archivo, no un directorio.", directory)
	}

	entries, err := os.ReadDir(path)
	if err != nil {
		return "Error al listar directorio: " + err.Error()
	}
	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })

	filesDir, _ := filepath.Abs(config.FilesDir)
	var items []string
	for _, e := range entries {
		fullPath := filepath.Join(path, e.Name())
		rel, _ := filepath.Rel(filesDir, fullPath)
		if e.IsDir() {
			sub, _ := os.ReadDir(fullPath)
			items = append(items, fmt.Sprintf("[DIR]  %s/  (%d elementos)", rel, len(sub)))
		} else {
			fi, _ := e.Info()
			items = append(items, fmt.Sprintf("[FILE] %s  (%s bytes)", rel, formatInt(fi.Size())))
		}
	}
	if len(items) == 0 {
		return "El directorio está vacío."
	}
	label := "/"
	if directory != "" {
		label = directory
	}
	return fmt.Sprintf("Contenido de %s  (%d elementos):\n%s", label, len(items), strings.Join(items, "\n"))
}

func DeleteFile(filename string) string {
	path, err := safePath(filename)
	if err != nil {
		return "Error: " + err.Error()
	}
	info, err := os.Stat(path)
	if err != nil {
		return fmt.Sprintf("Error: '%s' no existe.", filename)
	}
	if info.IsDir() {
		if err := os.RemoveAll(path); err != nil {
			return "Error al eliminar directorio: " + err.Error()
		}
		return fmt.Sprintf("Directorio '%s' eliminado correctamente.", filename)
	}
	if err := os.Remove(path); err != nil {
		return "Error al eliminar: " + err.Error()
	}
	return fmt.Sprintf("Archivo '%s' eliminado correctamente.", filename)
}

func SearchInFiles(query, directory string) string {
	if strings.TrimSpace(query) == "" {
		return "Error: La búsqueda no puede estar vacía."
	}
	var searchDir string
	var err error
	if directory == "" {
		searchDir, err = filepath.Abs(config.FilesDir)
	} else {
		searchDir, err = safePath(directory)
	}
	if err != nil {
		return "Error: " + err.Error()
	}
	if _, err := os.Stat(searchDir); err != nil {
		return "Error: El directorio de búsqueda no existe."
	}

	filesDir, _ := filepath.Abs(config.FilesDir)
	queryLower := strings.ToLower(query)
	var results []string
	filesSearched := 0

	filepath.WalkDir(searchDir, func(p string, d fs.DirEntry, err error) error {
		if err != nil || d.IsDir() {
			return nil
		}
		if !textExtensions[strings.ToLower(filepath.Ext(p))] {
			return nil
		}
		data, err := os.ReadFile(p)
		if err != nil {
			return nil
		}
		filesSearched++
		lines := strings.Split(string(data), "\n")
		var matches [][2]string
		for i, line := range lines {
			if strings.Contains(strings.ToLower(line), queryLower) {
				matches = append(matches, [2]string{fmt.Sprintf("%d", i+1), line})
			}
		}
		if len(matches) == 0 {
			return nil
		}
		rel, _ := filepath.Rel(filesDir, p)
		results = append(results, fmt.Sprintf("\n=== %s ===", rel))
		shown := matches
		if len(shown) > 5 {
			shown = shown[:5]
		}
		for _, m := range shown {
			line := m[1]
			if len(line) > 200 {
				line = line[:200]
			}
			results = append(results, fmt.Sprintf("  Línea %s: %s", m[0], strings.TrimSpace(line)))
		}
		if len(matches) > 5 {
			results = append(results, fmt.Sprintf("  ... y %d coincidencia(s) más", len(matches)-5))
		}
		return nil
	})

	if len(results) == 0 {
		return fmt.Sprintf("No se encontró '%s' en %d archivo(s) revisado(s).", query, filesSearched)
	}
	return fmt.Sprintf("Se encontró '%s' en %d archivo(s) (de %d revisados):%s",
		query, len(results), filesSearched, strings.Join(results, ""))
}

func GetFileInfo(filename string) string {
	path, err := safePath(filename)
	if err != nil {
		return "Error: " + err.Error()
	}
	info, err := os.Stat(path)
	if err != nil {
		return fmt.Sprintf("Error: '%s' no existe.", filename)
	}
	filesDir, _ := filepath.Abs(config.FilesDir)
	rel, _ := filepath.Rel(filesDir, path)
	fileType := "Archivo"
	if info.IsDir() {
		fileType = "Directorio"
	}
	lines := []string{
		fmt.Sprintf("Nombre    : %s", info.Name()),
		fmt.Sprintf("Ruta      : %s", rel),
		fmt.Sprintf("Tipo      : %s", fileType),
		fmt.Sprintf("Tamaño    : %s bytes (%.1f KB)", formatInt(info.Size()), float64(info.Size())/1024),
		fmt.Sprintf("Modificado: %s", info.ModTime().Format("2006-01-02 15:04:05")),
	}
	if info.IsDir() {
		entries, _ := os.ReadDir(path)
		lines = append(lines, fmt.Sprintf("Elementos : %d", len(entries)))
	} else if ext := filepath.Ext(info.Name()); ext != "" {
		lines = append(lines, fmt.Sprintf("Extensión : %s", ext))
	}
	return strings.Join(lines, "\n")
}

func FormatSize(bytes int64) string {
	switch {
	case bytes >= 1024*1024:
		return fmt.Sprintf("%.2f MB", float64(bytes)/(1024*1024))
	case bytes >= 1024:
		return fmt.Sprintf("%.1f KB", float64(bytes)/1024)
	default:
		return fmt.Sprintf("%d B", bytes)
	}
}

func formatInt(n int64) string {
	s := fmt.Sprintf("%d", n)
	var result []byte
	for i, c := range s {
		if i > 0 && (len(s)-i)%3 == 0 {
			result = append(result, ',')
		}
		result = append(result, byte(c))
	}
	return string(result)
}
